schematics/exceptions.py

```diff
@@ -26,49 +29,71 @@ def __init__(self, message, errors=None, *args):
         and therefore immutable. As a result, it is not possible to
         mutate BaseError's error list or dict after initialization.
         """
+        errors = self._freeze(errors)
         if message is None and errors is not None:
             message = str(self._to_primitive(errors))
         super(BaseError, self).__init__(message)
         self._errors = errors
 
     @property
     def errors(self):
-        return copy.deepcopy(self._errors)
+        return self._errors
 
     @property
     def messages(self):
         """ an alias for errors, provided for compatibility with V1. """
-        return self.errors
+        return self._errors
 
     def to_primitive(self):
         """
         converts the errors dict to a primitive representation of dicts,
         list and strings.
         """
-        return self._to_primitive(self._errors)
+        if not hasattr(self, "_primitive"):
+            self._primitive = self._to_primitive(self._errors)
+        return self._primitive
+
+    @staticmethod
+    def _freeze(obj):
```

Comment:
I wanted to note that it's still possible to modify elements and dictionary values of FrozenDict and FrozenList. At a performance cost, I can do a recursive _freeze to guard against this.